### PR TITLE
handle slashing from p2p

### PR DESCRIPTION
**Pull request by shayzluf**
Part of #4836

This PR adds handling of slashings from p2p network after slashing pool have been implemented

### CODE REVIEW

**Comment by shayzluf** [comment]
@terencechain seems like that is what we do here https://github.com/prysmaticlabs/prysm/blob/456ac5f9a3ac05c8860ea1435c9df6fddc79e3c4/beacon-chain/rpc/beacon/slashings.go#L19

**Comment by shayzluf** [comment]
@prestonvanloon added tests

File: beacon-chain/sync/subscriber_test.go [hunk 1]
```diff
@@ -7,6 +7,10 @@ import (
 	"testing"
 	"time"
 
+	"github.com/prysmaticlabs/prysm/beacon-chain/operations/slashings"
+
```
Comment (rauljordan) [inline beacon-chain/sync/subscriber_test.go]:
```suggestion
```

File: beacon-chain/sync/subscriber_test.go [hunk 2]
```diff
@@ -7,6 +7,10 @@ import (
...
+	"github.com/prysmaticlabs/prysm/shared/params"
+
```
Comment (rauljordan) [inline beacon-chain/sync/subscriber_test.go]:
```suggestion
```

Comment (shayzluf) [reply beacon-chain/sync/subscriber_test.go]: thanks

File: beacon-chain/sync/subscriber_test.go (hunk 1 above)
Comment (shayzluf) [reply beacon-chain/sync/subscriber_test.go]: removed

**Comment by terencechain** [comment]
> @terencechain seems like that is what we do here
> 
> https://github.com/prysmaticlabs/prysm/blob/456ac5f9a3ac05c8860ea1435c9df6fddc79e3c4/beacon-chain/rpc/beacon/slashings.go#L19

That's a bug. Just raised it in the slashing-detection channel.

File: beacon-chain/sync/subscriber_handlers.go [hunk 3]
```diff
@@ -17,11 +20,22 @@ func (r *Service) voluntaryExitSubscriber(ctx context.Context, msg proto.Message
 }
 
 func (r *Service) attesterSlashingSubscriber(ctx context.Context, msg proto.Message) error {
-	// TODO(#3259): Requires handlers in operations service to be implemented.
+	as := msg.(*ethpb.AttesterSlashing)
```
Comment (terencechain) [inline beacon-chain/sync/subscriber_handlers.go]: type assertion on `msg.(*ethpb.AttesterSlashing)` ?

defensive programming 101 😆

Comment (shayzluf) [reply beacon-chain/sync/subscriber_handlers.go]: i added it but then saw that the code that calls it is calling it by its type https://github.com/prysmaticlabs/prysm/blob/6d2a2ebadf984ec7c77e5b4450f4806148b92906/beacon-chain/p2p/gossip_topic_mappings.go#L12
i will add it anyway for resilience

Comment (terencechain) [reply beacon-chain/sync/subscriber_handlers.go]: Correct. Hence defensive programming

Comment (shayzluf) [reply beacon-chain/sync/subscriber_handlers.go]: thanks

File: beacon-chain/sync/subscriber_handlers.go [hunk 4]
```diff
@@ -2,26 +2,50 @@ package sync
...
 	return nil
 }
 
 func (r *Service) attesterSlashingSubscriber(ctx context.Context, msg proto.Message) error {
-	// TODO(#3259): Requires handlers in operations service to be implemented.
+	as, ok := msg.(*ethpb.AttesterSlashing)
+	if !ok {
+		return fmt.Errorf("wrong type, expected: *ethpb.AttesterSlashing got: %T", msg)
+	}
+	s, err := r.db.State(ctx, bytesutil.ToBytes32(as.Attestation_1.Data.BeaconBlockRoot))
```
Comment (rauljordan) [inline beacon-chain/sync/subscriber_handlers.go]: Somebody can spam panic nodes by sending us slashings that have these nil inner fields, fixing

**Comment by shayzluf** [comment]
@prestonvanloon can you please take another look ? added tests

File: beacon-chain/sync/subscriber_handlers.go [hunk 5]
```diff
@@ -2,26 +2,56 @@ package sync
...
 
 func (r *Service) attesterSlashingSubscriber(ctx context.Context, msg proto.Message) error {
-	// TODO(#3259): Requires handlers in operations service to be implemented.
+	as, ok := msg.(*ethpb.AttesterSlashing)
+	if !ok {
+		return fmt.Errorf("wrong type, expected: *ethpb.AttesterSlashing got: %T", msg)
+	}
+	// Do some nil checks to prevent easy DoS'ing of this handler.
+	if as != nil && as.Attestation_1 != nil && as.Attestation_1.Data != nil {
+		s, err := r.db.State(ctx, bytesutil.ToBytes32(as.Attestation_1.Data.BeaconBlockRoot))
```
Comment (prestonvanloon) [inline beacon-chain/sync/subscriber_handlers.go]: What if no state found?

**Pull request merged**